### ISSUE TITLE
nbFile and nbShell

### ISSUE THREAD

**Issue by pietroppeter** [pietroppeter/nimib#34]
`nbFile`: write nim code content into a file that might be compiled/run later
`nbShell`: run shell command and capture output and error

Example:

```
nbFile("module.nim"): # write the following content to module.nim file
  proc hello*(name: string) = echo "hello " & name

nbFile("main.nim"): # write the following content to main.nim file
  import module
  hello "you"

nbShell("nim r main") # run main.nim and captures output
```

this could also be used for creating nim files that can be compiled with a different backend (e.g. javascript). example from karax:

```
nbFile("hello.nim"):
  include karax / prelude
  
  proc createDom(): VNode =
    result = buildHtml(tdiv):
      text "Hello World!"
  
  setRenderer createDom

nbShell("karun -r helloworld.nim")
```

* in html output the code block should show the name of the file (possibly also a different styling?).
* nbShell should be styled differently from standard nim code block
* not sure what the advantage of using nim code instead of a string, since the nim code will not checked by tools like `nim check` or `nimsuggest` (basically code will be discarded and only its string representation will be kept).

**Comment by Clonkk** [comment]
The nbFile functionality is a great idea and would help me write the tutorial on interacting with Python and Julia. 

Are you already working on it ?

**Comment by pietroppeter** [comment]
No, not working on it. The idea for me was to finish the PR on refactoring NbBlock and rendering, but I have had really little time last few months (hopefully this situation is changing). I do not see any problem if you want to already give it an implementation (thanks!) and I will refactor that when the time comes.

**Comment by Clonkk** [comment]
Do you intend to keep the file in the book ?

**Comment by Clonkk** [comment]
I'll keep the file for now, so it's stays simple. It's easy enough to remove manually if need be

**Comment by pietroppeter** [comment]
realized now I never answered to this, sorry. Anyway, I agree behavior for nbFile should be to keep the file, we could at some point provide some more ergonomic system for cleaning up files.

**Comment by HugoGranstrom** [comment]
A use-case I've stumbled upon when trying some things with NimiSlide:

I have to write javascript code, event listeners specifically, and inline it in the outputted HTML file. I could write the javascript in a `nbRawOutput` but it's error-prone and I have to write javascript (!). So what I could do instead would be to use `nbFile` and `nbShell` to compile it to a file and read that file into the HTML. *But* now every time I create this block it will create a new file which it loads, which isn't very elegant and the different scripts can't really communicate with each other in a safe way. Because what I ultimately would like is for all these javascript snippets to have access to a global variable. In my case, it's an array of indices specifying which fragments contain this block and which doesn't. So every javascript snippet has to have access to this. My idea then is to add a new block that behaves like `nbFile` except it doesn't immediately save the content to file but rather accumulates the code from each of them (perhaps wrapping each one in a `block`) and then compiles all of them together into js. 

This way we could write the javascript in Nim plus have the type-safety of Nim instead of writing javascript in a string which is a pain to debug. 

Just saw that you removed the untyped `nbFile` though so perhaps this won't work out for the same reasons it didn't 😅 

And to clarify, my use-case is this:
```nim
var currentFragment = 0 # is incremented each fragment

nbGlobalJS:
	var fragment2id: Table[int, int]

nbSpecificBlock:
	nbRawOutput: """<p id="0"></p>"""
	nbLocalJS:
		fragment2id[currentFragment] = 0

nbSpecificBlock:
	nbRawOutput: """<p id="1"></p>"""
	nbLocalJS:
		fragment2id[currentFragment] = 1

####### Turns this into HTML:
```
```html
<p id="0"></p>
<p id="1"></p>
<script>
	let fragment2id = {}
	fragment2id[23] = 0
	fragment2id[30] = 1
</script>
```

When writing this example I also realized we need to be able to capture variables from the compiled code, `currentFragment` in this case.

This grew bigger than I imagined when I started writing 🤣 Does this seem like a desirable and possible feature?

**Comment by pietroppeter** [comment]
This looks like a use case I would definitely would want to support. 

As a possible api alternative to what you propose:

```nim
script = nbInitCode:
  var myVar = 0

script.addCode:
  inc myVar

script.addToDocAsJs
```

The above I think can be adapted to be use for your case.

Notes:
- `nbInitCode` would create a new block with code, without running code (not happy about the name, since nbCode is normally something where code is run; I would not like nbFile either though). This code block is not added to the document.
- `addCode`: adds code to block (again without running it)
- `addToDocAsJs` takes a block, writes and compile to js the code content (using a temp file), then adds the output in a `script` block to the document

Note that `nbFile` with untyped content was removed I think because I had issues when importing from it in the same file, but maybe it should be kept with some warning.

**Comment by HugoGranstrom** [comment]
Nice! :D And as usual you manage to simplify my ideas into more useful building blocks 😎 

I agree that `nbNewCode` isn't an ideal name for it, I would prefer a name that clearly states that it is a code block that isn't run and can be accumulated into. Although, `nbCode` could theoretically be implemented using this in combination with a `script.run()` proc so I wouldn't say it's inaccurate tbh.

The whole idea of assigning the script to a variable is great, then multiple scripts can be created more easily👍

> it would probably make sense in this context to add a nbJsCode block that does the above in one step (adds as js script the js compiled content of a single block), but it would not be what you are interested in for your use case.

Yes in the majority of cases this is all that is needed and it would be handy to have. 😄 I think the need for being able to capture variables is even more important in this case though as you may have to create an unique id for your tag in the compiled code and be able to access it in the JS but you have no global variables and as such can't do it all in the js. I have no idea how such a capture would be possible outside of using a macro or converting the code to string and manually replacing the variable with the value of the variable in the code. I'll have to give that a think-through.

> Note that nbFile with untyped content was removed I think because I had issues when importing from it in the same file, but maybe it should be kept with some warning.

I'm not really understanding what is importing and what causes the problem 😅 Being able to import `std/dom` is essential for this idea to work.

**Comment by pietroppeter** [comment]
> I think the need for being able to capture variables is even more important in this case

I do not actually understand what this "capture variables" means in this context and what is the use case, but it seems better to continue the discussion in the newly created issue.

> I'm not really understanding what is importing and what causes the problem

I have not been clear there, I was thinking of this (or similar) example of use case of `nbFile`:

```nim
nbFile("utils.nim"):
  var a* = 1

nbCode:
  from utils import a
```

it is indeed a contrived example and the `nbCode` block will not compile until the `nbFile` block is executed once. Honestly, now that I think about it I am not sure if this was the reason why I removed the untyped `nbFile` example, it seems to have no bearing here, but it was something along these lines. Anyway, I should probably put it back and maybe while writing a test for it I will be able to understand what is the issue (or I was wrong and there was no issue).
